src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -614,7 +614,23 @@ private async Task<List<ChainedHeaderBlock>> RewindToForkPointAsync(ChainedHeade
                     this.SetConsensusTipInternalLocked(current.Previous);
                 }
 
-                var disconnectedBlock = new ChainedHeaderBlock(current.Block, current);
+                Block block = current.Block;
+
+                if (block == null)
+                {
+                    this.logger.LogTrace("Block '{0}' wasn't cached. Loading it from the database.", current.HashBlock);
+                    block = await this.blockStore.GetBlockAsync(current.HashBlock).ConfigureAwait(false);
+
+                    if (block == null)
```

Comment:
What I am not sure about here is so we really need to throw in case the block is not found.
I know we said rewind will include the block, but it does not have to and it simplifies the code greatly.
Currently the only component that depends on a block being there is mempool rewind.

Reply:
We are throwing here just because it's a sanity check and should never happen. 

However I kinda see your point so moved this sanity check right before we need those blocks so we can see a clear error in case it ever happens.

Reply:
>Currently the only component that depends on a block being there is mempool rewind.

We potentially will need blocks being disconnected for other components in the future so it's worth it to have this logic.

Reply:
No the fix you had was good, my point was about requiring the block to be present in the signaler.

Reply:
>my point was about requiring the block to be present in the signaler.

Hm, ok. 
I think having the block always set is a good guarantee that other components need. Otherwise this signaler is not reliable and can't be used for different things.